In [76]:
import os
from dotenv import load_dotenv
load_dotenv()

True

In [77]:

GOOGLE_API_KEY=os.getenv("GOOGLE_API_KEY")
TAVILY_API_KEY=os.getenv("TAVILY_API_KEY")
GROQ_API_KEY=os.getenv("GROQ_API_KEY")
LANGCHAIN_API_KEY=os.getenv("LANGCHAIN_API_KEY")
LANGCHAIN_PROJECT=os.getenv("LANGCHAIN_PROJECT")

In [78]:
LANGCHAIN_PROJECT

'langgraph-prerequist'

In [79]:

os.environ["GOOGLE_API_KEY"] = GOOGLE_API_KEY
os.environ["TAVILY_API_KEY"] = TAVILY_API_KEY
os.environ["GROQ_API_KEY"]= GROQ_API_KEY
os.environ["LANGCHAIN_API_KEY"] = LANGCHAIN_API_KEY
os.environ["LANGCHAIN_TRACING_V2"] = "true"
os.environ["LANGCHAIN_ENDPOINT"] = "https://api.smith.langchain.com"
os.environ["LANGCHAIN_PROJECT"]=LANGCHAIN_PROJECT

In [7]:
pip install langchain-community


  Using cached dataclasses_json-0.6.7-py3-none-any.whl.metadata (25 kB)
  Using cached pydantic_settings-2.5.2-py3-none-any.whl.metadata (3.5 kB)
  Using cached marshmallow-3.22.0-py3-none-any.whl.metadata (7.2 kB)
  Using cached typing_inspect-0.9.0-py3-none-any.whl.metadata (1.5 kB)
  Using cached mypy_extensions-1.0.0-py3-none-any.whl.metadata (1.1 kB)
   ---------------------------------------- 0.0/2.4 MB ? eta -:--:--
   ------ --------------------------------- 0.4/2.4 MB 7.6 MB/s eta 0:00:01
   ------------ --------------------------- 0.7/2.4 MB 6.5 MB/s eta 0:00:01
   --------------- ------------------------ 0.9/2.4 MB 5.2 MB/s eta 0:00:01
   ------------------ --------------------- 1.1/2.4 MB 4.9 MB/s eta 0:00:01
   ----------------------- ---------------- 1.4/2.4 MB 4.8 MB/s eta 0:00:01
   -------------------------- ------------- 1.6/2.4 MB 4.2 MB/s eta 0:00:01
   -------------------------- ------------- 1.6/2.4 MB 4.2 MB/s eta 0:00:01
   -------------------------- ----------


[notice] A new release of pip is available: 23.3.2 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [8]:
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.document_loaders import WebBaseLoader
from langchain_community.vectorstores import Chroma

USER_AGENT environment variable not set, consider setting it to identify your requests.


In [9]:
from langchain_google_genai import GoogleGenerativeAIEmbeddings
embeddings = GoogleGenerativeAIEmbeddings(model="models/embedding-001")
from langchain_google_genai import ChatGoogleGenerativeAI
llm = ChatGoogleGenerativeAI(model="gemini-1.0-pro")

In [10]:
'''from langchain_huggingface import HuggingFaceEmbeddings
embeddings=HuggingFaceEmbeddings(model_name="all-MiniLM-L6-v2")'''
from langchain_groq import ChatGroq
import os
llm=ChatGroq(model_name="Gemma2-9b-It")

In [12]:
'''from langchain_huggingface import HuggingFaceEmbeddings
embeddings=HuggingFaceEmbeddings(model_name="all-MiniLM-L6-v2")'''
from langchain_groq import ChatGroq
import os
llm2=ChatGroq(model_name="Gemma2-9b-It")
llm2.invoke("tell me happy birthday song")

AIMessage(content="Happy birthday to you,\nHappy birthday to you,\nHappy birthday, dear [Your Name],\nHappy birthday to you! 🎂🎉 \n\n\nLet me know if you'd like me to sing it in a different tone or with some extra musical flair! 🎶  \n\n", additional_kwargs={}, response_metadata={'token_usage': {'completion_tokens': 61, 'prompt_tokens': 14, 'total_tokens': 75, 'completion_time': 0.110909091, 'prompt_time': 8.378e-05, 'queue_time': 0.014205579, 'total_time': 0.110992871}, 'model_name': 'Gemma2-9b-It', 'system_fingerprint': 'fp_10c08bf97d', 'finish_reason': 'stop', 'logprobs': None}, id='run-b1e86b0b-d53a-430f-bf11-fad43b00ec51-0', usage_metadata={'input_tokens': 14, 'output_tokens': 61, 'total_tokens': 75})

In [11]:
while True:
    question=input("type your question. if you want to quit the chat write quit")
    if question !="quit":
        print(llm.invoke(question).content)
    else:
        print("goodbye take care yourself")
        break

Hello! 👋 How can I help you today? 😊  

A black hole is a region of spacetime where gravity is so strong that nothing, not even light, can escape. 

Here's a breakdown:

* **Extreme Gravity:** Black holes form when massive stars collapse at the end of their life cycle. This collapse concentrates an immense amount of mass into an incredibly small space.
* **Event Horizon:** The boundary around a black hole beyond which escape is impossible is called the event horizon.  
* **Singularity:** At the center of a black hole is thought to be a singularity, a point of infinite density where the laws of physics as we know them break down.

**Key Points:**

* **Invisible:** We can't see black holes directly because they don't emit light.
* **Detected indirectly:** We detect them by observing their gravitational effects on nearby objects, such as stars or gas.
* **Types:** Black holes come in different sizes:
    * **Stellar-mass black holes:**  Form from the collapse of individual stars.
    * **

In [13]:
from langchain_core.chat_history import BaseChatMessageHistory
from langchain_core.chat_history import InMemoryChatMessageHistory
from langchain_core.runnables.history import RunnableWithMessageHistory
from langchain_core.messages import AIMessage

In [14]:
store={}

In [15]:

def get_session_history(session_id: str) -> BaseChatMessageHistory:
    if session_id not in store:
        store[session_id] = InMemoryChatMessageHistory()
    return store[session_id]

In [16]:

config = {"configurable": {"session_id": "firstchat"}}

In [17]:
model_with_memory=RunnableWithMessageHistory(llm,get_session_history)

In [19]:

model_with_memory.invoke(("Hi! I'm Paritosh Shukla"),config=config).content

"Hi Paritosh Shukla! \n\nIt's nice to meet you.  What can I do for you today? 😊 \n"

In [21]:
store

{'firstchat': InMemoryChatMessageHistory(messages=[HumanMessage(content="Hi! I'm Paritosh Shukla", additional_kwargs={}, response_metadata={}), AIMessage(content="Hi Parm Paritosh Shukla! Nice to meet you.  \n\nIt's great to have you here. Is there anything I can help you with today? 😊  \n\n", additional_kwargs={}, response_metadata={'token_usage': {'completion_tokens': 40, 'prompt_tokens': 18, 'total_tokens': 58, 'completion_time': 0.072727273, 'prompt_time': 8.224e-05, 'queue_time': 0.014257678, 'total_time': 0.072809513}, 'model_name': 'Gemma2-9b-It', 'system_fingerprint': 'fp_10c08bf97d', 'finish_reason': 'stop', 'logprobs': None}, id='run-db59d279-4cf4-4e85-8f0c-d394241f9f4b-0', usage_metadata={'input_tokens': 18, 'output_tokens': 40, 'total_tokens': 58}), HumanMessage(content="Hi! I'm Paritosh Shukla", additional_kwargs={}, response_metadata={}), AIMessage(content="Hi Paritosh Shukla! \n\nIt's nice to meet you.  What can I do for you today? 😊 \n", additional_kwargs={}, response_m

In [20]:
model_with_memory.invoke(("tell me what is my name?"),config=config).content

'Your name is Paritosh Shukla. 😊  \n\nI remember!  Is there anything else I can help you with?  \n'

In [32]:
from langchain_community.document_loaders import TextLoader, DirectoryLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.vectorstores import FAISS
from langchain import PromptTemplate
from langchain_core.runnables import RunnableParallel, RunnablePassthrough , RunnableLambda
from langchain_core.output_parsers import StrOutputParser

### Reading the txt files from source directory

loader = DirectoryLoader('C:\\Users\\ashutosh\\OneDrive\\Desktop\\Gen AI\\LangGraph\\data', glob="./*.txt", loader_cls=TextLoader)
docs = loader.load()

### Creating Chunks using RecursiveCharacterTextSplitter

text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=50,
    chunk_overlap=10,
    length_function=len
)
new_docs = text_splitter.split_documents(documents=docs)
doc_strings = [doc.page_content for doc in new_docs]

###  BGE Embddings
from langchain_google_genai import GoogleGenerativeAIEmbeddings
embeddings = GoogleGenerativeAIEmbeddings(model="models/embedding-001")
'''from langchain.embeddings import HuggingFaceBgeEmbeddings

model_name = "BAAI/bge-base-en-v1.5"
model_kwargs = {'device': 'cpu'}
encode_kwargs = {'normalize_embeddings': True} # set True to compute cosine similarity
embeddings = HuggingFaceBgeEmbeddings(
    model_name=model_name,
    model_kwargs=model_kwargs,
    encode_kwargs=encode_kwargs,
)
'''

### Creating Retriever using Vector DB

db = FAISS.from_documents(new_docs, embeddings)
retriever = db.as_retriever(search_kwargs={"k": 4})

In [31]:
pip install faiss-cpu

   ---------------------------------------- 0.0/14.9 MB ? eta -:--:--
   ---------------------------------------- 0.0/14.9 MB ? eta -:--:--
   ---------------------------------------- 0.0/14.9 MB 325.1 kB/s eta 0:00:46
   ---------------------------------------- 0.0/14.9 MB 325.1 kB/s eta 0:00:46
   ---------------------------------------- 0.0/14.9 MB 325.1 kB/s eta 0:00:46
   ---------------------------------------- 0.0/14.9 MB 325.1 kB/s eta 0:00:46
   ---------------------------------------- 0.0/14.9 MB 325.1 kB/s eta 0:00:46
   ---------------------------------------- 0.1/14.9 MB 379.3 kB/s eta 0:00:39
   - -------------------------------------- 0.5/14.9 MB 1.4 MB/s eta 0:00:11
   -- ------------------------------------- 1.0/14.9 MB 2.5 MB/s eta 0:00:06
   --- ------------------------------------ 1.4/14.9 MB 3.4 MB/s eta 0:00:04
   ----- ---------------------------------- 1.9/14.9 MB 3.9 MB/s eta 0:00:04
   ------ --------------------------------- 2.4/14.9 MB 4.5 MB/s eta 0:00:03
 


[notice] A new release of pip is available: 23.3.2 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [39]:
template = """Answer the question based only on the following context:
{context}
and ellaborate the answer in short 2 3 line
Question: {question}
"""
prompt = PromptTemplate.from_template(template)


In [40]:

retrieval_chain = (
    RunnableParallel({"context": retriever, "question": RunnablePassthrough()})
    | prompt
    | llm
    | StrOutputParser()
    )

In [41]:
question =" What are the best crops to grow in sandy soil?"
print(retrieval_chain.invoke(question))

The provided context states: "Sandy soil is well-draining, making it ideal".  

While it doesn't list specific crops, well-draining soil is generally good for crops that tolerate drought conditions and don't like "wet feet". 



# Let's Start with Tools and Agents

In [42]:
from langchain_community.tools import WikipediaQueryRun
from langchain_community.utilities import WikipediaAPIWrapper

In [44]:
pip install wikipedia

  Using cached wikipedia-1.4.0.tar.gz (27 kB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
  Created wheel for wikipedia: filename=wikipedia-1.4.0-py3-none-any.whl size=11705 sha256=8f08734017443229b3cf55f272eeb8f6ff3f176368ad65de7d8d5230b3d49e98
  Stored in directory: c:\users\ashutosh\appdata\local\pip\cache\wheels\63\47\7c\a9688349aa74d228ce0a9023229c6c0ac52ca2a40fe87679b8
Successfully built wikipedia
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.3.2 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [45]:
api_wrapper = WikipediaAPIWrapper(top_k_results=5, doc_content_chars_max=100)

In [46]:
tool = WikipediaQueryRun(api_wrapper=api_wrapper)

In [47]:
tool.name

'wikipedia'

In [48]:
tool.description

'A wrapper around Wikipedia. Useful for when you need to answer general questions about people, places, companies, facts, historical events, or other subjects. Input should be a search query.'

In [49]:
tool.args

{'query': {'description': 'query to look up on wikipedia',
  'title': 'Query',
  'type': 'string'}}

In [50]:
tool.return_direct

False

In [51]:
print(tool.run({"query": "langchain"}))

Page: LangChain
Summary: LangChain is a software framework that helps facilitate the integration of 


In [52]:
tool.run("langchain")

'Page: LangChain\nSummary: LangChain is a software framework that helps facilitate the integration of '

In [53]:
from langchain_core.pydantic_v1 import BaseModel, Field

C:\Users\ashutosh\AppData\Roaming\Python\Python312\site-packages\IPython\core\interactiveshell.py:3526: LangChainDeprecationWarning: As of langchain-core 0.3.0, LangChain uses pydantic v2 internally. The langchain_core.pydantic_v1 module was a compatibility shim for pydantic v1, and should no longer be used. Please update the code to import from Pydantic directly.

For example, replace imports like: `from langchain_core.pydantic_v1 import BaseModel`
with: `from pydantic import BaseModel`
or the v1 compatibility namespace if you are working in a code base that has not been fully upgraded to pydantic 2 yet. 	from pydantic.v1 import BaseModel

  exec(code_obj, self.user_global_ns, self.user_ns)


In [54]:
class WikiInputs(BaseModel):
    query: str = Field(description="query to look up in Wikipedia, should be 3 or less words")

In [55]:
tool=WikipediaQueryRun(
    name="wiki-tool",
    description="look up things in wikipedia",
    args_schema=WikiInputs,
    api_wrapper=api_wrapper,
    return_direct=True,
)

ValidationError: 1 validation error for WikipediaQueryRun
args_schema
  Input should be a subclass of BaseModel [type=is_subclass_of, input_value=<class '__main__.WikiInputs'>, input_type=ModelMetaclass]
    For further information visit https://errors.pydantic.dev/2.9/v/is_subclass_of

In [56]:
tool.name

'wikipedia'

In [57]:
tool.description

'A wrapper around Wikipedia. Useful for when you need to answer general questions about people, places, companies, facts, historical events, or other subjects. Input should be a search query.'

In [58]:
tool.args

{'query': {'description': 'query to look up on wikipedia',
  'title': 'Query',
  'type': 'string'}}

In [59]:
tool.return_direct

False

In [60]:
print(tool.run("langchain"))

Page: LangChain
Summary: LangChain is a software framework that helps facilitate the integration of 


# youtube search tool

In [61]:
from langchain_community.tools import YouTubeSearchTool

In [62]:
tool=YouTubeSearchTool()

In [63]:
tool.name

'youtube_search'

In [64]:
tool.description

'search for youtube videos associated with a person. the input to this tool should be a comma separated list, the first part contains a person name and the second a number that is the maximum number of video results to return aka num_results. the second part is optional'

In [66]:
pip install youtube_search

  Using cached youtube_search-2.1.2-py3-none-any.whl.metadata (1.2 kB)
Using cached youtube_search-2.1.2-py3-none-any.whl (3.4 kB)
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.3.2 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [68]:
tool.run("Indias Got Latend")

"['https://www.youtube.com/watch?v=aQpJ7OgXVn8&pp=ygURSW5kaWFzIEdvdCBMYXRlbmQ%3D', 'https://www.youtube.com/watch?v=Vsx9a7-IXX4&pp=ygURSW5kaWFzIEdvdCBMYXRlbmQ%3D']"

In [80]:
from langchain_community.tools.tavily_search import TavilySearchResults

In [81]:
# tavily_api_key=

In [82]:
tool = TavilySearchResults()

In [83]:
tool.invoke({"query": "What happened in the latest burning man floods"})

[{'url': 'https://www.npr.org/2023/09/03/1197497458/the-latest-on-the-burning-man-flooding',
  'content': "There are also reports that at least one person has died at the counterculture festival about a hundred miles north of Reno, Nev. Earlier this afternoon, I caught up with NPR's Claudia Peschiutta, who's at her first burn, and she told me it's muddy where she is, but that she and her camp family have been making the best of things.\n National\nThe latest on the Burning Man flooding\nClaudia Peschiutta\nAuthorities are investigating a death at the Burning Man festival in the Nevada desert after tens of thousands of people are stuck in camps because of rain.\n SCOTT DETROW, HOST:\nKnee-deep mud, warnings to conserve food and water, orders to shelter in place - this is all at Burning Man 2023 after torrential rains turned the Black Rock Desert into miles and miles of mud. I mean, mostly what I've seen from my personal experience is just any sort of need that you have, somebody, whethe

In [84]:
from langchain import hub

In [85]:
from langchain.agents import AgentExecutor,create_openai_functions_agent

In [98]:
instructions = """You are an assistant."""
base_prompt = hub.pull("langchain-ai/openai-functions-template")
prompt = base_prompt.partial(instructions=instructions)

In [99]:
tavily_tool = TavilySearchResults()

In [100]:
tools = [tavily_tool]

In [101]:
agent = create_openai_functions_agent(llm, tools, prompt)

In [102]:
agent_executor = AgentExecutor(
    agent=agent,
    tools=tools,
    verbose=True,
)

In [103]:
print(agent_executor.invoke({"input": "who was the mahatma gandhi"}))



> Entering new AgentExecutor chain...


> Finished chain.
{'input': 'who was the mahatma gandhi', 'output': ''}


# Showcase that one more agent also

# Create our custom agent and custom tools

In [92]:
# Import things that are needed generically
from langchain.pydantic_v1 import BaseModel, Field
from langchain.tools import BaseTool, StructuredTool, tool
from pydantic import BaseModel

C:\Users\ashutosh\AppData\Roaming\Python\Python312\site-packages\IPython\core\interactiveshell.py:3526: LangChainDeprecationWarning: As of langchain-core 0.3.0, LangChain uses pydantic v2 internally. The langchain.pydantic_v1 module was a compatibility shim for pydantic v1, and should no longer be used. Please update the code to import from Pydantic directly.

For example, replace imports like: `from langchain.pydantic_v1 import BaseModel`
with: `from pydantic import BaseModel`
or the v1 compatibility namespace if you are working in a code base that has not been fully upgraded to pydantic 2 yet. 	from pydantic.v1 import BaseModel

  exec(code_obj, self.user_global_ns, self.user_ns)


In [93]:
@tool
def search(query: str) -> str:
    """Look up things online."""
    return "LangChain"

In [94]:
print(search.name)
print(search.description)
print(search.args)

search
Look up things online.
{'query': {'title': 'Query', 'type': 'string'}}


In [95]:
@tool
def multiply(a: int, b: int) -> int:
    """Multiply two numbers."""
    return a * b

In [96]:
print(multiply.name)
print(multiply.description)
print(multiply.args)

multiply
Multiply two numbers.
{'a': {'title': 'A', 'type': 'integer'}, 'b': {'title': 'B', 'type': 'integer'}}


In [97]:
class SearchInput(BaseModel):
    query: str = Field(description="should be a search query")

In [77]:

@tool("search-tool", args_schema=SearchInput, return_direct=True)
def search(query: str) -> str:
    """Look up things online."""
    return "LangChain"

In [78]:
print(search.name)
print(search.description)
print(search.args)
print(search.return_direct)

search-tool
Look up things online.
{'query': {'title': 'Query', 'type': 'string'}}
True


c:\Users\sunny\LangGraph-End-to-End-Course\venv\lib\site-packages\pydantic\json_schema.py:2191: PydanticJsonSchemaWarning: Default value default=PydanticUndefined description='should be a search query' extra={} is not JSON serializable; excluding default from JSON schema [non-serializable-default]
  warnings.warn(message, PydanticJsonSchemaWarning)


In [ ]:
from typing import Optional, Type

from langchain.callbacks.manager import (
    AsyncCallbackManagerForToolRun,
    CallbackManagerForToolRun,
)

In [79]:
class SearchInput(BaseModel):
    query: str = Field(description="should be a search query")

In [80]:
class CalculatorInput(BaseModel):
    a: int = Field(description="first number")
    b: int = Field(description="second number")

# here is my custom tool

In [82]:
from langchain.agents import tool
@tool
def get_word_length(word: str) -> int:
    """Returns the length of a word."""
    return len(word)

In [83]:
get_word_length.invoke("abc")

3

In [84]:
tools = [get_word_length]

In [85]:
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder

prompt = ChatPromptTemplate.from_messages(
    [
        (
            "system",
            "You are very powerful assistant, but don't know current events",
        ),
        ("user", "{input}"),
        MessagesPlaceholder(variable_name="agent_scratchpad"),
    ]
)

In [86]:
llm_with_tools = llm.bind_tools(tools)

In [88]:
from langchain.agents.format_scratchpad.openai_tools import (
    format_to_openai_tool_messages,
)
from langchain.agents.output_parsers.openai_tools import OpenAIToolsAgentOutputParser



In [89]:
agent = (
    {
        "input": lambda x: x["input"],
        "agent_scratchpad": lambda x: format_to_openai_tool_messages(
            x["intermediate_steps"]
        ),
    }
    | prompt
    | llm_with_tools
    | OpenAIToolsAgentOutputParser()
)

In [90]:
from langchain.agents import AgentExecutor

agent_executor = AgentExecutor(agent=agent, tools=tools, verbose=True)

In [91]:
list(agent_executor.stream({"input": "How many letters in the word eudca"}))



> Entering new AgentExecutor chain...

Invoking: `get_word_length` with `{'word': 'eudca'}`


55

> Finished chain.


[{'actions': [ToolAgentAction(tool='get_word_length', tool_input={'word': 'eudca'}, log="\nInvoking: `get_word_length` with `{'word': 'eudca'}`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'tool_calls': [{'index': 0, 'id': 'call_4188', 'function': {'arguments': '{"word":"eudca"}', 'name': 'get_word_length'}, 'type': 'function'}]}, response_metadata={'finish_reason': 'tool_calls'}, id='run-6cc153ac-50f7-468e-871a-bb81fd15b1f7', tool_calls=[{'name': 'get_word_length', 'args': {'word': 'eudca'}, 'id': 'call_4188', 'type': 'tool_call'}], usage_metadata={'input_tokens': 966, 'output_tokens': 86, 'total_tokens': 1052}, tool_call_chunks=[{'name': 'get_word_length', 'args': '{"word":"eudca"}', 'id': 'call_4188', 'index': 0, 'type': 'tool_call_chunk'}])], tool_call_id='call_4188')],
  'messages': [AIMessageChunk(content='', additional_kwargs={'tool_calls': [{'index': 0, 'id': 'call_4188', 'function': {'arguments': '{"word":"eudca"}', 'name': 'get_word_length'}, 'type': 'fu

[{'actions': [ToolAgentAction(tool='get_word_length', tool_input={'word': 'eudca'}, log="\nInvoking: `get_word_length` with `{'word': 'eudca'}`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'tool_calls': [{'index': 0, 'id': 'call_4188', 'function': {'arguments': '{"word":"eudca"}', 'name': 'get_word_length'}, 'type': 'function'}]}, response_metadata={'finish_reason': 'tool_calls'}, id='run-6cc153ac-50f7-468e-871a-bb81fd15b1f7', tool_calls=[{'name': 'get_word_length', 'args': {'word': 'eudca'}, 'id': 'call_4188', 'type': 'tool_call'}], usage_metadata={'input_tokens': 966, 'output_tokens': 86, 'total_tokens': 1052}, tool_call_chunks=[{'name': 'get_word_length', 'args': '{"word":"eudca"}', 'id': 'call_4188', 'index': 0, 'type': 'tool_call_chunk'}])], tool_call_id='call_4188')],
  'messages': [AIMessageChunk(content='', additional_kwargs={'tool_calls': [{'index': 0, 'id': 'call_4188', 'function': {'arguments': '{"word":"eudca"}', 'name': 'get_word_length'}, 'type': 'function'}]}, response_metadata={'finish_reason': 'tool_calls'}, id='run-6cc153ac-50f7-468e-871a-bb81fd15b1f7', tool_calls=[{'name': 'get_word_length', 'args': {'word': 'eudca'}, 'id': 'call_4188', 'type': 'tool_call'}], usage_metadata={'input_tokens': 966, 'output_tokens': 86, 'total_tokens': 1052}, tool_call_chunks=[{'name': 'get_word_length', 'args': '{"word":"eudca"}', 'id': 'call_4188', 'index': 0, 'type': 'tool_call_chunk'}])]},
 {'steps': [AgentStep(action=ToolAgentAction(tool='get_word_length', tool_input={'word': 'eudca'}, log="\nInvoking: `get_word_length` with `{'word': 'eudca'}`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'tool_calls': [{'index': 0, 'id': 'call_4188', 'function': {'arguments': '{"word":"eudca"}', 'name': 'get_word_length'}, 'type': 'function'}]}, response_metadata={'finish_reason': 'tool_calls'}, id='run-6cc153ac-50f7-468e-871a-bb81fd15b1f7', tool_calls=[{'name': 'get_word_length', 'args': {'word': 'eudca'}, 'id': 'call_4188', 'type': 'tool_call'}], usage_metadata={'input_tokens': 966, 'output_tokens': 86, 'total_tokens': 1052}, tool_call_chunks=[{'name': 'get_word_length', 'args': '{"word":"eudca"}', 'id': 'call_4188', 'index': 0, 'type': 'tool_call_chunk'}])], tool_call_id='call_4188'), observation=5)],
  'messages': [FunctionMessage(content='5', additional_kwargs={}, response_metadata={}, name='get_word_length')]},
 {'output': '5',
  'messages': [AIMessage(content='5', additional_kwargs={}, response_metadata={})]}]

In [92]:
llm.invoke("How many letters in the word educa")

AIMessage(content='Here\'s how to solve this:\n\n* **Focus on the phrase:**  The question asks about the letters in the phrase "letters in the word educa".\n* **Count carefully:** Count each letter in that phrase.\n\nLet me know if you\'d like me to do the counting for you! \n', additional_kwargs={}, response_metadata={'token_usage': {'completion_tokens': 67, 'prompt_tokens': 16, 'total_tokens': 83, 'completion_time': 0.121818182, 'prompt_time': 0.00024042, 'queue_time': 0.042495864, 'total_time': 0.122058602}, 'model_name': 'Gemma2-9b-It', 'system_fingerprint': 'fp_10c08bf97d', 'finish_reason': 'stop', 'logprobs': None}, id='run-ed352c53-84e7-48d6-9576-0acd99befd1f-0', usage_metadata={'input_tokens': 16, 'output_tokens': 67, 'total_tokens': 83})

In [93]:
from langchain_core.prompts import MessagesPlaceholder

MEMORY_KEY = "chat_history"
prompt = ChatPromptTemplate.from_messages(
    [
        (
            "system",
            "You are very powerful assistant, but bad at calculating lengths of words.",
        ),
        MessagesPlaceholder(variable_name=MEMORY_KEY),
        ("user", "{input}"),
        MessagesPlaceholder(variable_name="agent_scratchpad"),
    ]
)

In [94]:
from langchain_core.messages import AIMessage, HumanMessage

chat_history = []

In [95]:
agent = (
    {
        "input": lambda x: x["input"],
        "agent_scratchpad": lambda x: format_to_openai_tool_messages(
            x["intermediate_steps"]
        ),
        "chat_history": lambda x: x["chat_history"],
    }
    | prompt
    | llm_with_tools
    | OpenAIToolsAgentOutputParser()
)

In [96]:
agent_executor = AgentExecutor(agent=agent, tools=tools, verbose=True)

In [97]:
input1 = "how many letters in the word educa?"
result = agent_executor.invoke({"input": input1, "chat_history": chat_history})
chat_history.extend(
    [
        HumanMessage(content=input1),
        AIMessage(content=result["output"]),
    ]
)
agent_executor.invoke({"input": "is that a real word?", "chat_history": chat_history})

Error in StdOutCallbackHandler.on_chain_start callback: AttributeError("'NoneType' object has no attribute 'get'")



Invoking: `get_word_length` with `{'word': 'educa'}`


5

Error in StdOutCallbackHandler.on_chain_start callback: AttributeError("'NoneType' object has no attribute 'get'")


5

> Finished chain.
Yes. 


> Finished chain.


{'input': 'is that a real word?',
 'chat_history': [HumanMessage(content='how many letters in the word educa?', additional_kwargs={}, response_metadata={}),
  AIMessage(content='5', additional_kwargs={}, response_metadata={})],
 'output': 'Yes. \n'}